src/cookbook/gradual-composition-api-migration.md

```diff
@@ -0,0 +1,245 @@
+
+The Composition API is very flexible, but you may lose that implicit declaration in communication to fellow maintainers. Our suggestion is that if you do use it as a centralized state management store, that you place it in a `stateManagement` folder, or something similarly named, so that responsibilities are clear.
+
+## Components
+
+It is possible at this point to start to move your components to the Composition API, however, for larger code bases, you may not see immediate need or good reason to, as the Options API works nicely.
+
+Our suggestion is to use this opportunity to think through the architecture and refactor any larger components that have too much responsibility on their own.
+
+For instance, if you have a component that has the following methods:
+
+- one that gathers input from a form
+- one that validates that input
+- one that uses the input to call out to an API
+- one that does something with that API response
+
+You may find that you can potentially create a reusable component for the API call, as on a large application there's a fairly high chance you will need to make another.
+
+For the purposes of demonstration, here's how that one method refactor might look.
+
+### Before, with a Tightly Coupled Method
+
+```js
+methods: {
+  async getGitHubProjects() {
+    try {
+      this.isLoading = true
+      let projects = await fetch(
+        `https://api.github.com/users/sdras/repos?page=1&per_page=100`
+      )
+      this.gitHubProjects = await projects.json()
+    } catch (error) {
+      this.error = error
+      console.error(error);
+    } finally {
+      this.isLoading = false;
+    }
+  }
+}
+```
+
+### After, with the Composition API
+
+```js
+import { reactive, toRefs } from '@vue/composition-api'
+
+export function useFetchAPI(api, options) {
+  const state = reactive({
+    response: null,
+    isLoading: false,
+    hasErrors: null
+  })
+
+  const apiCall = async () => {
+    state.isLoading = true
+    try {
+      let res = await fetch(api, options)
+      state.response = await res.json()
+    } catch (error) {
+      state.hasErrors = error
+    } finally {
+      state.isLoading = false
+    }
+  }
+
+  return {
+    ...toRefs(state),
+    apiCall
+  }
+}
+```
+
+And now we can refactor our earlier component like this:
+
+```js
+import { useFetchAPI } from '@/features/useFetchAPI.js'
+
+export default {
+  data() {
+    return {
+      gitHubData: [],
+      errors: null,
+      loading: false
+    }
+  },
+  setup() {
+    const { hasErrors, isLoading, response, apiCall } = useFetchAPI(
+      `https://api.github.com/users/sdras/repos?page=1&per_page=100`,
+      {}
+    )
+
+    return {
+      hasErrors,
+      isLoading,
+      response,
+      apiCall
+    }
+  },
+  methods: {
+    async callGitHub() {
+      await this.apiCall()
+      this.gitHubData = this.response
+      this.errors = this.hasErrors
+      this.loading = this.isLoading
```

Comment:
As currently written, the `loading` flag is never set to `true`. `isLoading` does change to `true` but it isn't copied across to `loading` until the awaited promise has resolved, by which time it's back to `false`.

I'm guessing this example is going to be tweaked anyway, so this problem may well disappear as part of that.